model/atmosphere/advection/src/icon4py/model/atmosphere/advection/advection.py

```diff
@@ -0,0 +1,425 @@
+import dataclasses
+import enum
+import logging
+
+from icon4py.model.atmosphere.advection import (
+    advection_states,
+    advection_horizontal,
+    advection_vertical,
+)
+from icon4py.model.atmosphere.advection.stencils import (
+    apply_density_increment,
+    apply_interpolated_tracer_time_tendency,
+    copy_cell_kdim_field,
+)
+from icon4py.model.common import (
+    dimension as dims,
+    field_type_aliases as fa,
+    type_alias as ta,
+)
+from icon4py.model.common.decomposition import definitions as decomposition
+from icon4py.model.common.grid import horizontal as h_grid, icon as icon_grid, geometry
+from icon4py.model.common.utils import gt4py_field_allocation as field_alloc
+
+
+"""
+Advection module ported from ICON mo_advection_stepping.f90.
+"""
+
+# flake8: noqa
+log = logging.getLogger(__name__)
+
+
+class HorizontalAdvectionType(enum.Enum):
+    """
+    Horizontal operator scheme for advection.
+    """
+
+    #: no horizontal advection
+    NO_ADVECTION = 0
+    #: 2nd order MIURA with linear reconstruction
+    LINEAR_2ND_ORDER = 2
+
+
+class HorizontalAdvectionLimiter(enum.Enum):
+    """
+    Limiter for horizontal advection operator.
+    """
+
+    #: no horizontal limiter
+    NO_LIMITER = 0
+    #: positive definite horizontal limiter
+    POSITIVE_DEFINITE = 4
+
+
+class VerticalAdvectionType(enum.Enum):
+    """
+    Vertical operator scheme for advection.
+    """
+
+    #: no vertical advection
+    NO_ADVECTION = 0
+
+
+class VerticalAdvectionLimiter(enum.Enum):
+    """
+    Limiter for vertical advection operator.
+    """
+
+    #: no vertical limiter
+    NO_LIMITER = 0
+
+
+@dataclasses.dataclass(frozen=True)
+class AdvectionConfig:
+    horizontal_advection_type: HorizontalAdvectionType
+    horizontal_advection_limiter: HorizontalAdvectionLimiter
+    vertical_advection_type: VerticalAdvectionType
+    vertical_advection_limiter: VerticalAdvectionLimiter
+
+    """
+    Contains necessary parameters to configure an advection run.
+    """
+
+    def __post_init__(self):
+        self._validate()
+
+    def _validate(self):
+        """Apply consistency checks and validation on configuration parameters."""
+
+        assert (
+            self.horizontal_advection_type not in HorizontalAdvectionType.__members__
+        ), f"Horizontal advection type {self.horizontal_advection_type} not implemented."
+        assert (
+            self.horizontal_advection_limiter not in HorizontalAdvectionLimiter.__members__
+        ), f"Horizontal advection limiter {self.horizontal_advection_limiter} not implemented."
+        assert (
+            self.vertical_advection_type not in VerticalAdvectionType.__members__
+        ), f"Vertical advection type {self.vertical_advection_type} not implemented."
+        assert (
+            self.vertical_advection_limiter not in VerticalAdvectionLimiter.__members__
+        ), f"Vertical advection limiter {self.vertical_advection_limiter} not implemented."
```

Comment:
Since there are many asserts in this function, I would change them to if statements with return errors